lock-keeper/src/types/database/secrets.rs

```
@@ -1,18 +1,25 @@
 //! Database models for secrets
 
-use crate::crypto::{Encrypted, KeyId, Secret};
-
+use crate::crypto::{Encrypted, KeyId, Secret, SigningKeyPair};
 use serde::{Deserialize, Serialize};
 
+/// Holds user's stored secrets of all types
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct StoredSecrets {
+    pub arbitrary_secrets: Vec<StoredEncryptedSecret>,
+    pub signing_keys: Vec<StoredEncryptedSigningKeyPair>,
+    pub server_generated_signing_keys: Vec<StoredSigningKeyPair>,
+}
```

thought: This structure seems like every time we add a new type of stored secret, we have to update our retrieve code to look through another list. I know there's a plan to discuss better database structure, but is it a worthwhile intermediate step to pull out the key ids from this so it's easier to search on them? Like

```rust
// don't use this name
pub struct AnyKindOfSecret {
    key_id: KeyId
    retrieved: bool,
    secret: StoredKind,
}
enum StoredKind {
    EncryptedSecret(Encrypted<Secret>),
    EncryptedSigningKeyPair(Encrypted<SigningKeyPair>),
    SigningKeyPair(SigningKeyPair),
}
```

---

We avoided the enum because we don't want to add error handling for situations where you need an `EncryptedSecret` but you have to make sure it's not the other variants. Having it this way means that the types are always correct for the secret you're trying to retrieve. If the database data gets corrupted and the data for a given key id changes to a different type, the whole user will fail to load instead of having a silent failure that doesn't show up until the key is retrieved.

This struct is temporary until we have a more permanent solution for how we represent secrets in the database.

---

Alternatively we could remove the guarantee that the caller gets what they asked for and force them to check the type.

---

Also wanted to throw in there that because of the types, retrieving each type of secret has its own function right now, so we can specify which one of the lists to look through in each.

---

Okay, cool. That sounds good to me. 

> Alternatively we could remove the guarantee that the caller gets what they asked for and force them to check the type.

This seems more mistake-prone. What you have planned makes sense to me.